[Cocoa] Add support for experimental showCaptionDisplaySettings() API

jernoble opened the pull request:
#### 79fb1f414fe35c868ee79a0c859d31bf29fab720
<pre>
[Cocoa] Add support for experimental showCaptionDisplaySettings() API
<a href="https://rdar.apple.com/164744943">rdar://164744943</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=302547">https://bugs.webkit.org/show_bug.cgi?id=302547</a>

Reviewed by Eric Carlson.

In <a href="https://github.com/WebKit/explainers/tree/main/CaptionDisplaySettings">https://github.com/WebKit/explainers/tree/main/CaptionDisplaySettings</a>, WebKit proposed
a new API which allows web apps to trigger a system-provided UI for picking subtitle
display styles. This patch adds initial support for that API.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/CaptionDisplaySettingsOptions.h: Copied from Source/WebKit/UIProcess/PageClient.cpp.
* Source/WebCore/html/CaptionDisplaySettingsOptions.idl: Copied from Source/WebKit/UIProcess/PageClient.cpp.
* Source/WebCore/html/HTMLVideoElement+CaptionDisplaySettings.idl: Copied from Source/WebKit/UIProcess/PageClient.cpp.
* Source/WebCore/html/HTMLVideoElementCaptionDisplaySettings.cpp: Added.
(WebCore::parsePositionAreaString):
(WebCore::HTMLVideoElementCaptionDisplaySettings::showCaptionDisplaySettings):
* Source/WebCore/html/HTMLVideoElementCaptionDisplaySettings.h: Copied from Source/WebKit/UIProcess/PageClient.cpp.
* Source/WebCore/page/ChromeClient.cpp:
(WebCore::ChromeClient::showCaptionDisplaySettings):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::showCaptionDisplaySettings): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/PageClient.cpp:
(WebKit::PageClient::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::showCaptionDisplaySettings): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant _resetMediaControlsContextMenuPresenter]):
(-[WKActionSheetAssistant captionStyleMenuDidClose:]):
(-[WKActionSheetAssistant showCaptionDisplaySettingsMenu:withOptions:completionHandler:]):
(-[WKActionSheetAssistant contextMenuInteraction:willEndForConfiguration:animator:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView showCaptionDisplaySettingsMenu:withOptions:completionHandler:]):
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerIOS.mm:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::showCaptionDisplaySettings):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::showCaptionDisplaySettings):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

Canonical link: <a href="https://commits.webkit.org/303334@main">https://commits.webkit.org/303334@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b9ae82f599a35bc489b4231bd44cfc38ba38282

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132009 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4501 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43025 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139523 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83914 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2e3ba73f-e593-40a2-9e61-f9dda49009c8) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133879 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4444 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4263 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100912 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cb212d96-ba77-4c77-bcfd-23f573dade5d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134955 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3167 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118230 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81702 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/72a253a0-c158-46b0-b1a8-44f62eb17955) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3062 "Passed tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82741 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111816 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36349 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142168 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4171 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36931 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109280 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4252 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3636 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109452 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27736 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3160 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114507 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57392 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4224 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32902 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4056 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67671 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4316 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4184 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -9042,3 +9042,24 @@ struct WebCore::FocusOptions {
     bool preventScroll;
     std::optional<bool> focusVisible;
 };
+
+header: <WebCore/ResolvedCaptionDisplaySettingsOptions.h>
+[CustomHeader] enum class WebCore::ResolvedCaptionDisplaySettingsOptionsXPositionArea : uint8_t {
+    Center,
+    Left,
+    Right,
+};
+
+header: <WebCore/ResolvedCaptionDisplaySettingsOptions.h>
+[CustomHeader] enum class WebCore::ResolvedCaptionDisplaySettingsOptionsYPositionArea : uint8_t {
+    Center,
+    Top,
+    Bottom,
+};
+
+header: <WebCore/ResolvedCaptionDisplaySettingsOptions.h>
+[CustomHeader] struct WebCore::ResolvedCaptionDisplaySettingsOptions {
+    std::optional<WebCore::FloatRect> anchorBounds;
+    std::optional<WebCore::ResolvedCaptionDisplaySettingsOptionsXPositionArea> xPositionArea;
+    std::optional<WebCore::ResolvedCaptionDisplaySettingsOptionsYPositionArea> yPositionArea;
+};
```
Nit: I think the convention is to create a new `.serializer.in` file for new structs and classes

jernoble committed:
[Cocoa] Add support for experimental showCaptionDisplaySettings() API
rdar://164744943
https://bugs.webkit.org/show_bug.cgi?id=302547

Reviewed by Eric Carlson.

In https://github.com/WebKit/explainers/tree/main/CaptionDisplaySettings, WebKit proposed
a new API which allows web apps to trigger a system-provided UI for picking subtitle
display styles. This patch adds initial support for that API.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/CaptionDisplaySettingsOptions.h: Copied from Source/WebKit/UIProcess/PageClient.cpp.
* Source/WebCore/html/CaptionDisplaySettingsOptions.idl: Copied from Source/WebKit/UIProcess/PageClient.cpp.
* Source/WebCore/html/HTMLVideoElement+CaptionDisplaySettings.idl: Copied from Source/WebKit/UIProcess/PageClient.cpp.
* Source/WebCore/html/HTMLVideoElementCaptionDisplaySettings.cpp: Added.
(WebCore::parsePositionAreaString):
(WebCore::HTMLVideoElementCaptionDisplaySettings::showCaptionDisplaySettings):
* Source/WebCore/html/HTMLVideoElementCaptionDisplaySettings.h: Copied from Source/WebKit/UIProcess/PageClient.cpp.
* Source/WebCore/page/ChromeClient.cpp:
(WebCore::ChromeClient::showCaptionDisplaySettings):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::showCaptionDisplaySettings): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/PageClient.cpp:
(WebKit::PageClient::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::showCaptionDisplaySettings): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant _resetMediaControlsContextMenuPresenter]):
(-[WKActionSheetAssistant captionStyleMenuDidClose:]):
(-[WKActionSheetAssistant showCaptionDisplaySettingsMenu:withOptions:completionHandler:]):
(-[WKActionSheetAssistant contextMenuInteraction:willEndForConfiguration:animator:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView showCaptionDisplaySettingsMenu:withOptions:completionHandler:]):
* Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerIOS.mm:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::showCaptionDisplaySettings):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::showCaptionDisplaySettings):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::showCaptionDisplaySettings):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

Canonical link: https://commits.webkit.org/303334@main
